bessctl/measurement_utils.py

```diff
@@ -0,0 +1,393 @@
+        agg = dict()
+        for k in old:
+            agg[k] = (old[k] + new[k]) / 2
+        return agg
+
+    def next(self):
+        throughput = self.bess.get_port_stats(self.tx_port)
+        if self.tx_port != self.rx_port:
+            rx_stats = self.bess.get_port_stats(self.rx_port)
+            throughput.inc.packets = rx_stats.inc.packets
+            throughput.inc.dropped = rx_stats.inc.dropped
+            throughput.inc.bytes = rx_stats.inc.bytes
+
+        rtt, jitter = None, None
+        if self.measure:
+            arg = {'clear': True,
+                   'latency_percentiles': self.rtt_percentiles,
+                   'jitter_percentiles': self.rtt_percentiles}
+            mstats = self.bess.run_module_command(self.measure, 'get_summary',
+                                                  'MeasureCommandGetSummaryArg',
+                                                  arg)
+            rtt = dict()
+            jitter = dict()
+            for i, p in enumerate(self.rtt_percentiles):
+                rtt[p] = mstats.latency.percentile_values_ns[i]
+                jitter[p] = mstats.jitter.percentile_values_ns[i]
+
+        if self.rate:
+            now = time.time()
+            delta_t = now - self.last_check
+            if self.old_throughput == None:
+                self.old_throughput = copy.copy(throughput)
+                self.old_throughput.inc.packets = 0
+                self.old_throughput.inc.dropped = 0
+                self.old_throughput.inc.bytes = 0
+                self.old_throughput.out.packets = 0
+                self.old_throughput.out.dropped = 0
+                self.old_throughput.out.bytes = 0
+
+            delta_throughput = PortStatsGenerator.throughput_delta(
+                self.old_throughput, throughput, delta_t)
+            self.old_throughput = throughput
+            self.last_check = now
+
+            agg_rtt, agg_jitter = None, None
+            if self.measure:
+                agg_rtt = PortStatsGenerator.aggregate_rtt(self.old_rtt, rtt)
+                agg_jitter = PortStatsGenerator.aggregate_rtt(
+                    self.old_jitter, jitter)
+                self.old_rtt = rtt
+                self.old_jitter = jitter
+
+            return PortStats(delta_throughput, agg_rtt, agg_jitter)
+
+        return PortStats(throughput, rtt, jitter)
+
+
+class PortConfig(object):
+
+    def __init__(self, name, num_queues=1, pci=None, port_id=None, vdev=None,
+                 no_tx=False, no_rx=False):
+        """
+        Creates a configuraiton for a Port named `name` and configures it with `num_queues` tx and
+        rx queues. If `no_tx` is set, no PortOut will be created. If `no_rx` is
+        set, no PortInc will be created. If `bess` is None, will attempt to
+        create all ports and modules on bessd running on localhost.
+        """
+        if sum(map(int, [pci is not None, port_id is not None, vdev is not None])) != 1:
+            raise Exception(
```

Comment:
Probably should be `TypeError` (that seems to be the most typical variant on "bad arguments").

Reply:
Fixed